wa/workloads/speedometer/__init__.py

```diff
@@ -12,57 +12,221 @@
+                webserver_cmd, shell=True, stdout=dev_null, stderr=dev_null
+            )
+
+        subprocess.check_output(
+            "adb reverse tcp:{0} tcp:{0}".format(self.webserver_port), shell=True
+        )
+
+        # Clear the cache...
+        self.target.execute("pm clear {}".format(self.chrome_package, as_root=True))
+
+        # Launch the browser for the first time...
+        browser_launch_cmd = "am start -a android.intent.action.VIEW -d {} {}".format(
+            self.speedometer_url, self.chrome_package
+        )
+
+        self.target.execute(browser_launch_cmd)
+        time.sleep(1)
+        self.target.execute("am force-stop {}".format(self.chrome_package))
+        time.sleep(1)
+
+        # Pull the preferences file from the device, modify it, and push it back.
+        # This is done to bypass the 'first launch' screen of the browser.
+        self.preferences_xml = "{}_preferences.xml".format(self.chrome_package)
+
+        self.target.pull(
+            "/data/data/{}/shared_prefs/{}".format(
+                self.chrome_package, self.preferences_xml
+            ),
+            context.output_directory,
+            as_root=True,
+        )
+
+        with open(
+            os.path.join(context.output_directory, self.preferences_xml)
+        ) as read_fh:
+            lines = read_fh.readlines()
+            lines.insert(
+                len(lines) - 1, '<boolean name="first_run_flow" value="true" />\n'
+            )
+            lines.insert(
+                len(lines) - 1,
+                '<boolean name="first_run_tos_accepted" value="true" />\n',
+            )
+            lines.insert(
+                len(lines) - 1,
+                '<boolean name="first_run_signin_complete" value="true" />\n',
+            )
+            with open(
+                os.path.join(context.output_directory, self.preferences_xml + ".new"),
+                "w",
+            ) as write_fh:
+                for line in lines:
+                    write_fh.write(line)
+
+        self.target.push(
+            os.path.join(context.output_directory, self.preferences_xml + ".new"),
+            "/data/data/{}/shared_prefs/{}".format(
+                self.chrome_package, self.preferences_xml
+            ),
+            as_root=True,
+        )
+
+    def run(self, context):
+        super(Speedometer, self).run(context)
+
+        browser_launch_cmd = "am start -a android.intent.action.VIEW -d {} {}".format(
+            self.speedometer_url, self.chrome_package
+        )
+        self.target.execute(browser_launch_cmd)
+
+        # Wait 60 seconds at least, and then wait until we don't see the 'sandboxed_process' process for 10 seconds.
+        time.sleep(60)
+
+        countdown = 5
+        while countdown > 0:
+            busiest_line = self.target.execute("top -n1 -m1 -q -b").split("\n")[0]
+            while "sandboxed_process" in busiest_line:
+                countdown = 5
```

Comment:
I don't understand the logic of setting the countdown inside the inner loop here. And this inner loop doesn't seem to have a break condition in case "standboxed_process" never goes away. 

This whole nested loop method seems unnecessarily convoluted. Can this be simplified/done via a different method? At the _very_ least,  leave a comment describing wtf is going on here.

Reply:
I've liberally commented this now. :) If 'sandboxed_process' never stops being the hottest process in the system (something my comments/changes will hopefully make clearer is what I'm looking for), then we should indeed not break out of the inner loop, because the benchmark is still running.

Reply:
oh my....